### PR TITLE
Add a table function to view persisted records

### PR DESCRIPTION
**Pull request by justinj**
A little loose at the moment since I think it's a bit of a pain to
thread through the configured persistence directory, but should enable
some amount of testing. I tried to factor out the behaviour here as much
as I could but it was a bit of a pain. I'm a bit unsure about the
package boundaries in general here.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4259)

### CODE REVIEW

**Comment by benesch** [comment]
I'm not sure I have the full context here. What's the goal? To be able to test persistence via the SQL interface?

**Comment by justinj** [comment]
Yes, that's the goal, to be able to write tests without polling/scanning the directory, it wound up being a bit more complex than I anticipated so it's perhaps worth revisiting if it's worth the effort.

**Comment by benesch** [comment]
Adding the table function seems fine to me, but I'm kind of wondering whether these persistence tests should just be normal Rust integration tests that are internal to the persistence code? No strong thoughts either way! I'll leave the full review to Ruchir since I don't have a ton of context.

File: src/dataflow/src/source/persistence.rs [hunk 1]
```diff
@@ -31,106 +30,6 @@ static RECORD_FILE_PREFIX: &str = "materialize";
...
-            Some(record)
-        } else {
-            None
-        }
-    }
-}
-
 /// Describes what is provided from a persisted file.
 #[derive(Debug)]
 pub struct RecordFileMetadata {
```
Comment (ruchirK) [inline src/dataflow/src/source/persistence.rs:35]: I think everything about `RecordFileMetadata` should eventually just end up wherever the `Record` type ends up (I think the only things the `dataflow crate needs to keep is the `PersistenceSender` type alias and the `WorkerPersistenceData` struct)

File: src/expr/src/record_iter.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
+// Copyright Materialize, Inc. All rights reserved.
```
Comment (ruchirK) [inline src/expr/src/record_iter.rs:1]: I wonder if everything in this file makes more sense as a file in the `repr` crate given that its the persisted file representation of our rows / datums? wdyt justin? i also think having all of that in one crate makes it nicer to test

File: src/expr/src/record_iter.rs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    }
+}
+
+impl RecordIter {
+    pub fn new(data: Vec<u8>) -> Self {
+        RecordIter { data, offset: 0 }
+    }
+
+    // TODO(justin): this should return an error.
+    pub fn files_for_source(id: GlobalId) -> Vec<PathBuf> {
```
Comment (ruchirK) [inline src/expr/src/record_iter.rs:125]: I'm a little bit confused why this is a method on `RecordIter` and not just a free function?

File: src/expr/src/record_iter.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                value: value.into(),
+            },
+            offset + len + 4,
+        ))
+    }
+}
+
+/// Iterator through a persisted set of records.
+#[derive(Debug)]
+pub struct RecordIter {
```
Comment (ruchirK) [inline src/expr/src/record_iter.rs:99]: can we call this PersistedRecordIter now that the underlying type is a `PersistedRecord`?

File: src/expr/src/record_iter.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        RecordIter { data, offset: 0 }
+    }
+
+    // TODO(justin): this should return an error.
+    pub fn files_for_source(id: GlobalId) -> Vec<PathBuf> {
+        // TODO(justin): plumb through the configured persistence directory, for now this is only
+        // suitable for testing.
+        let persistence_dir = PathBuf::from("mzdata/persistence");
+        let source_path = persistence_dir.join(id.to_string());
+        match std::fs::read_dir(&source_path) {
```
Comment (ruchirK) [inline src/expr/src/record_iter.rs:130]: I think this has a lot in common with whats inside `augment_connector_inner` -- can we maybe take whats here + some of the friendlier error handling from that function and then reuse it inside `augment_connector_inner`

**Comment by ruchirK** [comment]
> Adding the table function seems fine to me, but I'm kind of wondering whether these persistence tests should just be normal Rust integration tests that are internal to the persistence code? No strong thoughts either way! I'll leave the full review to Ruchir since I don't have a ton of context.

I think having it be integration tests is kind of hard because the persistence behavior spans the sql, dataflow, and coord crates.

I also think the table functions could be useful for customers to introspect what has been persisted? its a bit unfortunate that we can only give them binary blobs atm though :(

**Review by ruchirK — APPROVED** [review]
LGTM!

I left some ideas on how we might reorganize stuff that previously used to live in the dataflow crate into the repr crate - I think the persistence.rs file in the dataflow crate can probably be retired now

My only other comment is -- did we try using this table function in a scenario where Materialize has persisted 10 - 15 GB of data? how does it perform in that case? im less worried about latencies and more worried about potential oom-kills

**Review by justinj — COMMENTED** [review]
I think it will not perform well in that case, but I think for the current purpose of testing this is the best we can do, unless you have an idea for how to handle it?

> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4259#-:-MI-i4gL0jkxuOtLUYam:bj92m42)* status: 0 of 10 files reviewed, 5 unresolved discussions (waiting on @benesch and @ruchirK)

___
*[src/dataflow/src/source/persistence.rs, line 35 at r1](https://reviewable.io/reviews/materializeinc/materialize/4259#-MHm32CcSAihaxpPAP8D-r1-35:-MI-hYZmDzRP9Pp5B4zk:b-61mfnb) ([raw file](https://github.com/materializeinc/materialize/blob/df447618afa407a848ee1a5b4918dc414ba45ec5/src/dataflow/src/source/persistence.rs#L35)):*
<details><summary><i>Previously, ruchirK (Ruchir Khaitan) wrote…</i></summary><blockquote>

I think everything about `RecordFileMetadata` should eventually just end up wherever the `Record` type ends up (I think the only things the `dataflow` crate needs to keep is the `PersistenceSender` type alias and the `WorkerPersistenceData` struct)
</blockquote></details>

This is hard because `RecordFileMetadata` needs to see `GlobalId`, so it can't live in `repr`.
___
*[src/expr/src/record_iter.rs, line 1 at r1](https://reviewable.io/reviews/materializeinc/materialize/4259#-MHm32CcSAihaxpPAP8G-r1-1:-MI-iRLlD6QauvusYOQW:bp49qdg) ([raw file](https://github.com/materializeinc/materialize/blob/df447618afa407a848ee1a5b4918dc414ba45ec5/src/expr/src/record_iter.rs#L1)):*
<details><summary><i>Previously, ruchirK (Ruchir Khaitan) wrote…</i></summary><blockquote>

I wonder if everything in this file makes more sense as a file in the `repr` crate given that its the persisted file representation of our rows / datums? wdyt justin? i also think having all of that in one crate makes it nicer to test if / when we add unit tests
</blockquote></details>

I moved it to `repr`
___
*[src/expr/src/record_iter.rs, line 99 at r1](https://reviewable.io/reviews/materializeinc/materialize/4259#-MHm32CcSAihaxpPAP8G-r1-99:-MI-iTeh2GzcrkQ6k8Tm:b-896fix) ([raw file](https://github.com/materializeinc/materialize/blob/df447618afa407a848ee1a5b4918dc414ba45ec5/src/expr/src/record_iter.rs#L99)):*
<details><summary><i>Previously, ruchirK (Ruchir Khaitan) wrote…</i></summary><blockquote>

can we call this PersistedRecordIter now that the underlying type is a `PersistedRecord`?
</blockquote></details>

Done.
___
*[src/expr/src/record_iter.rs, line 125 at r1](https://reviewable.io/reviews/materializeinc/materialize/4259#-MHm32CcSAihaxpPAP8G-r1-125:-MI-iU4p9e6g0BmzmgHE:b-k1larm) ([raw file](https://github.com/materializeinc/materialize/blob/df447618afa407a848ee1a5b4918dc414ba45ec5/src/expr/src/record_iter.rs#L125)):*
<details><summary><i>Previously, ruchirK (Ruchir Khaitan) wrote…</i></summary><blockquote>

I'm a little bit confused why this is a method on `RecordIter` and not just a free function?
</blockquote></details>

It's an associated function, not a method, it just seemed reasonable to namespace it here? I pulled it out.
___
*[src/expr/src/record_iter.rs, line 130 at r1](https://reviewable.io/reviews/materializeinc/materialize/4259#-MHm32CcSAihaxpPAP8G-r1-130:-MI-kGLs1rnv5ocb4_4K:b-b63u98) ([raw file](https://github.com/materializeinc/materialize/blob/df447618afa407a848ee1a5b4918dc414ba45ec5/src/expr/src/record_iter.rs#L130)):*
<details><summary><i>Previously, ruchirK (Ruchir Khaitan) wrote…</i></summary><blockquote>

I think this has a lot in common with whats inside `augment_connector_inner` -- can we maybe take whats here + some of the friendlier error handling from that function and then reuse it inside `augment_connector_inner`
</blockquote></details>

I don't really see how to combine them in a nice way, can you help me understand what you're thinking?

**Review by ruchirK — APPROVED** [review]
I think thats totally fair for now

> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4259#-:-MI0MkHdAXQOMxXNxAa_:b-t419bk)* status: 0 of 10 files reviewed, 5 unresolved discussions (waiting on @benesch, @justinj, and @ruchirK)

___
*[src/expr/src/record_iter.rs, line 130 at r1](https://reviewable.io/reviews/materializeinc/materialize/4259#-MHm32CcSAihaxpPAP8G-r1-130:-MI0dp2HBoxapsB3TpaO:bt47uru) ([raw file](https://github.com/materializeinc/materialize/blob/df447618afa407a848ee1a5b4918dc414ba45ec5/src/expr/src/record_iter.rs#L130)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I don't really see how to combine them in a nice way, can you help me understand what you're thinking?
</blockquote></details>

yeah the crate distribution of everything is confusing. i think this is fine for now / I will think harder about how to refactor things after this merges (I think we can remove the dataflow crate's dependency on RecordFileMetadata, and then we can sort out where persistence stuff needs to live between coord, expr, and repr)

**Pull request merged**